pkg/controller/import-controller.go

```diff
@@ -139,15 +146,95 @@ func NewImportController(mgr manager.Manager, log logr.Logger, importerImage, pu
 	if err != nil {
 		return nil, err
 	}
-	if err := addImportControllerWatches(mgr, importController); err != nil {
+	if err := addImportControllerWatches(
+		mgr,
+		importController,
+		client,
+		log,
+	); err != nil {
 		return nil, err
 	}
 	return importController, nil
 }
 
-func addImportControllerWatches(mgr manager.Manager, importController controller.Controller) error {
+func addImportControllerWatches(mgr manager.Manager, importController controller.Controller, c client.Client, log logr.Logger) error {
 	// Setup watches
-	if err := importController.Watch(&source.Kind{Type: &corev1.PersistentVolumeClaim{}}, &handler.EnqueueRequestForObject{}); err != nil {
+
+	if err := importController.Watch(
+		&source.Kind{Type: &corev1.PersistentVolumeClaim{}},
+		handler.Funcs{
+			CreateFunc: func(evt event.CreateEvent, q workqueue.RateLimitingInterface) {
+				if evt.Meta == nil {
+					log.Error(nil, "CreateFunc received with no metadata", "event", evt)
+					return
+				}
+				q.Add(reconcile.Request{NamespacedName: types.NamespacedName{
+					Name:      evt.Meta.GetName(),
+					Namespace: evt.Meta.GetNamespace(),
+				}})
+			},
+			UpdateFunc: func(evt event.UpdateEvent, q workqueue.RateLimitingInterface) {
+				if evt.MetaOld != nil {
+					q.Add(reconcile.Request{NamespacedName: types.NamespacedName{
+						Name:      evt.MetaOld.GetName(),
+						Namespace: evt.MetaOld.GetNamespace(),
+					}})
+				} else {
+					log.Error(nil, "UpdateEvent received with no old metadata", "event", evt)
+				}
+
+				if evt.MetaNew != nil {
+					q.Add(reconcile.Request{NamespacedName: types.NamespacedName{
+						Name:      evt.MetaNew.GetName(),
+						Namespace: evt.MetaNew.GetNamespace(),
+					}})
+				} else {
+					log.Error(nil, "UpdateEvent received with no new metadata", "event", evt)
+				}
+			},
+			DeleteFunc: func(evt event.DeleteEvent, q workqueue.RateLimitingInterface) {
```

Comment:
I am afraid you are right, this means I need to go back re-implement this

Reply:
Ok I've learned a little bit about the CDI controller. So in case of missing the delete callback:

- If the controller restarts it will get the `main` PVC event from the UpdateFunc on first cache synchronization - it can reconcile this correctly. 
- If it just returns error during reconciliation, then I suppose the Reconcile will be retried. 
- If all else fails (event never delivered - is that possible?), then the next re-synchronization will Reconcile the PVC. But the re-synchronization for our cache runs every 10 hours by default.

Reply:
Anyway if you still think using a finalizer is more robust then I need to think how to handle this. I can put finalizer while creating the scratch pvc. Then there are a few scenarios possible:

1. Pod is Running, scratchPVC is Bound - the finalizer has to be removed, so it does not block normal deletion. Everything is ok.

2. Scratch was created and someone executed delete on it - before it was bound (but why), it has finalizer so it is marked for termination but still exists!. Pod is Pending but cannot correctly start: `Warning  FailedMount       2m32s (x3 over 5m15s)  kubelet, node02    Unable to attach or mount volumes: unmounted volumes=[cdi-scratch-vol], unattached volumes=[cdi-scratch-vol default-token-4qsdb cdi-data-vol]: error processing PVC cdi-e2e-tests-dv-func-test-dr5kp/scratch-space-delete-scratch: PVC is being deleted`. 

So here the new logic would need to remove the finalizer so PVC can be correctly Deleted, and then recreate the scratchPVC. If create PVC fails (because the old one still exists) it can return error and can be retried.

Is that ok?

Reply:
@mhenriks what do you think

Reply:
Okay, if reconciling the target PVC will address the issue, then finalizer is not necessary.  

I would suggest getting rid of these custom handler funcs, though.  Add an annotation to the scratch PVC that has the target PVC name.  When reconciling a PVC, look for that annotation and if it exists, get that PVC and reconcile it.

Reply:
I agree, it seems simpler and more explicit than following owner of an owner

Reply:
Ok, so I have added a new annotation and simplified the custom handler func. But the func was not removed as it is the only way to access the previous scratch PVC and schedule a PVC for it.  In the reconcile loop on the delete event the scratch PVC is just missing - disappears.

So considering how rare the original problem might be (someone or something trying to delete scratch space just after creation), and that now I cannot recreate the problem in tests. this is in my opinion a good enough solution. 

As mentioned in one of my previous comments, the only problem might be:
"If the DELETE event is never delivered, then the next re-synchronization will Reconcile the PVC, and we need to wait for re-sync."

Reply:
changed the main idea, removed the custom handler, now the Reconcile will be rescheduled again and again until pvc is complete